src/util/timelock.rs

```diff
@@ -0,0 +1,236 @@
+// Rust Bitcoin Library
+// Written in 2022 by
+//     Tobin C. Harding <me@tobin.cc>
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Bitcoin timelocks.
+//!
+//! Relative and absolute timelocks based on `nSequence` and `nLockTime`.
+//!
+
+#[cfg(feature = "std")]
+use std::error;
+
+use core::fmt;
+
+use blockdata::constants::{
+    LOCKTIME_THRESHOLD, SEQUENCE_LOCKTIME_DISABLE_FLAG, SEQUENCE_LOCKTIME_GRANULARITY,
+    SEQUENCE_LOCKTIME_TYPE_FLAG,
+};
+
+/// An absolute time lock, expires after either height or time.
+///
+/// This is nLockTime as defined in BIP-65.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Abs(u32);
+
+impl Abs {
+    /// Returns true if this absolute lock is lock-by-blockheight.
+    pub fn is_block_time(self) -> bool {
+        self.0 < LOCKTIME_THRESHOLD
+    }
+
+    /// Returns true if this absolute lock is lock-by-blocktime.
+    pub fn is_block_height(self) -> bool {
+        self.0 >= LOCKTIME_THRESHOLD
+    }
+
+    /// Returns true if this timelock is the same type as `other` timelock.
+    pub fn is_same_type(&self, other: Self) -> bool {
+        self.is_block_time() && other.is_block_time()
+            || self.is_block_height() && other.is_block_height()
+    }
+
+    /// Returns true if the timelock is locked at `age`.
+    pub fn is_locked(&self, age: u32) -> Result<bool, Error> {
+        self.is_expired(age).map(|b| !b)
+    }
+
+    /// Returns true if the timelock has expired at `age`.
+    pub fn is_expired(&self, age: u32) -> Result<bool, Error> {
+        let age = Abs::from(age);
+        if !self.is_same_type(age) {
+            return Err(Error::InvalidComparison);
+        }
+        Ok(age.0 >= self.0)
+    }
+
+    /// Returns true if this timelock value is zero.
+    pub fn is_zero(&self) -> bool {
+        self.0 == 0
+    }
+
+    /// Returns the nLockTime value. This is the inverse of `From<u32>`.
+    ///
+    /// # Examples
+    /// ```
+    /// # use bitcoin::timelock::Abs;
+    /// let n = 0xdeadbeef;
+    /// let lock = Abs::from(n);
+    /// assert_eq!(lock.n_lock_time(), n);
+    ///```
+    pub fn n_lock_time(&self) -> u32 {
+        self.to_u32()
+    }
+
+    /// Returns the time lock value. This is the inverse of `From<u32>`.
+    // FIXME: Should we provide this method as well as n_lock_time() or just one of them?
```

Comment:
IMHO we should have `to_u32` but not `n_lock_time`.

Reply:
Done. Removed `timelock::Rel::n_sequence` as well in favour of `to_u32`.